src/main/java/gg/projecteden/nexus/features/shops/ResourceWorldCommand.java

```diff
@@ -0,0 +1,491 @@
+	public void onBlockDropItem(BlockDropItemEvent event) {
+		if (!isResourceWorld(event.getBlock()))
+			return;
+
+		event.getItems().removeIf(item ->
+			trySell(event.getPlayer(), event.getBlockState(), item.getItemStack()));
+	}
+
+	@EventHandler
+	public void onEntityExplode(EntityExplodeEvent event) {
+		if (!isResourceWorld(event.getLocation()))
+			return;
+
+		if (!(event.getEntity() instanceof TNTPrimed tnt))
+			return;
+
+		if (!(tnt.getSource() instanceof Player player))
+			return;
+
+		final Iterator<Block> iterator = event.blockList().iterator();
+		while (iterator.hasNext()) {
+			final Block next = iterator.next();
+			for (ItemStack drop : next.getDrops())
+				if (trySell(player, next.getState(), drop)) {
+					next.setType(Material.AIR);
+					iterator.remove();
+				}
+		}
+	}
+
+	private boolean trySell(Player player, BlockState block, ItemStack drop) {
+		if (isNullOrAir(block.getType()) || isNullOrAir(drop))
+			return false;
+
+		final Shop shopper = new ShopService().get(player);
+		final boolean disabled = shopper.getDisabledResourceMarketItems().contains(drop.getType());
+		if (disabled)
+			return false;
+
+		if (getLogger(block.getWorld()).contains(block.getLocation()))
+			return false;
+
+		return trySell(player, drop);
+	}
+
+	private static final Map<UUID, BigDecimal> profit = new HashMap<>();
+	private static final Map<UUID, Integer> taskIds = new HashMap<>();
+
+	private void actionBar(Player player) {
+		final UUID uuid = player.getUniqueId();
+		final BigDecimal number = profit.get(uuid);
+
+		if (number.signum() != 0) {
+			Tasks.cancel(taskIds.getOrDefault(uuid, -1));
+			ActionBarUtils.sendActionBar(player, "&a+" + prettyMoney(number));
+			final int taskId = Tasks.wait(Time.SECOND.x(3.5), () -> profit.put(uuid, new BigDecimal(0)));
+			taskIds.put(uuid, taskId);
+		}
+	}
+
+	private boolean trySell(Player player, ItemStack item) {
+		final Optional<Product> product = getMatchingProduct(item);
+		if (product.isEmpty())
+			return false;
+
+		if (!(product.get().getExchange() instanceof BuyExchange exchange))
+			throw new InvalidInputException("Cannot process resource market exchange: " + camelCase(product.get().getExchangeType()));
+
+		process(player, item, exchange);
+		actionBar(player);
+
+		return true;
+	}
+
+	private void process(Player player, ItemStack item, BuyExchange exchange) {
+		BigDecimal current = profit.getOrDefault(player.getUniqueId(), new BigDecimal(0));
+		profit.put(player.getUniqueId(), current.add(exchange.processResourceMarket(player, item)));
+	}
+
+	private Optional<Product> getMatchingProduct(ItemStack item) {
+		return getResourceWorldProducts().stream()
+			.filter(product -> product.getItem().isSimilar(item))
+			.findFirst();
+	}
+
+	private List<Product> getResourceWorldProducts() {
+		return new ShopService().getMarket().getProducts().stream()
+			.filter(product -> product.isResourceWorld() && product.getExchangeType() == ExchangeType.BUY)
+			.toList();
+	}
```

Comment:
could be worth it to cache this value for a few seconds? not sure how expensive streams are, paper has led me to believe they are the devil but they're probably not bad enough to warrant implementing it, just an idea

Reply:
See above